File: scripts/sweep_tokenizer.mjs
#!/usr/bin/env node
import { readFileSync } from 'fs';
import { encode } from '../src/encoder.js';
import { trainTokenizer, tokenize, detokenize } from '../src/tokenizer/index.js';
import { encode as gptEncode } from 'gpt-tokenizer';

const TESTS = [
  ['Common English', 'the quick brown fox jumps over the lazy dog this is a test of the emergency broadcast system how now brown cow the rain in spain stays mainly on the plain'],
  ['Code', 'const app = express(); app.get("/api/tasks", async (req, res) => { const tasks = await db.query("SELECT * FROM tasks"); res.json(tasks); });'],
  ['SQL', "SELECT * FROM users WHERE id=42 AND name LIKE '%test%' ORDER BY created_at DESC; INSERT INTO table_name (a,b,c) VALUES (1,'x',true); UPDATE users SET name='abc', score=99 WHERE id=7; DELETE FROM users WHERE id>1000; CREATE TABLE test(id INT PRIMARY KEY,name VARCHAR(255));"],
  ['API', '{"status": "success", "data": {"users": [{"id": 1, "name": "John", "email": "john@example.com"}, {"id": 2, "name": "Jane", "email": "jane@example.com"}], "total": 2, "page": 1, "per_page": 10}}'],
  ['Shell', '$ echo "Hello, World!"; printf \'%s\\n\' "$HOME"; ls -la /tmp | grep ".log" && cat file.txt; sudo -n true || echo "no sudo"; python3 -c \'print("test")\'; node -e "console.log(42)"; git status --short; git add . && git commit -m "test"; npm run build && npm start'],
  ['Markdown', '# README.md ## Test Project ### Features - fast - simple - random ### Code `npm install && npm run dev` **bold** *italic* [link](https://example.com) > quote --- ### End'],
  ['Paths', 'C:\\Users\\Test\\file.txt D:\\Games\\MC\\server.exe /usr/bin/bash ~/.config/hypr/hyprland.conf ../../src/main.js ./build/output.log https://example.com/?a=1&b=2 ftp://x@y.z:21/path git@host:user/repo.git user@example.com test+tag@example.org'],
  ['Prompt', 'aicl is Goated BTW, and this can reduce tokens very vary fast'],
];

// Use a compact corpus: 50k chars from diverse sources
const rawCorpus = readFileSync('corpus/aicl_train.txt', 'utf-8');
const lines = rawCorpus.split('\n');
// Take every 25th line → ~1400 lines, ~50k chars
const trainCorpus = lines.filter((_, i) => i % 25 === 0).join('\n');
console.log(`Train corpus: ${trainCorpus.length} chars, ${trainCorpus.split('\n').length} lines`);

const testAicl = TESTS.map(([name, text]) => {
  const enc = encode(text);
  return { name, raw: text, aicl: enc.output, rawLen: [...text].length, aiclLen: [...enc.output].length };
});

const MERGE_COUNTS = [128, 256, 384, 512];
const results = [];

for (const numMerges of MERGE_COUNTS) {
  console.log(`\nTraining ${numMerges} merges...`);
  const t0 = Date.now();
  const vocab = trainTokenizer([trainCorpus], { numMerges, mergeBase: 100000, maxTokenLength: 5, minFrequency: 2 });
  const trainMs = Date.now() - t0;

  let totalAicl = 0, totalTokens = 0, totalGpt = 0, wins = 0;
  const rows = [];
  for (const t of testAicl) {
    const ids = tokenize(t.aicl, vocab);
    const ok = detokenize(ids, vocab) === t.aicl;
    const gpt = gptEncode(t.raw, { model: 'gpt-4o' }).length;
    const w = gpt / ids.length;
    totalAicl += t.aiclLen; totalTokens += ids.length; totalGpt += gpt;
    if (w >= 1.0) wins++;
    rows.push({ name: t.name, aicl: t.aiclLen, tok: ids.length, cpt: (t.aiclLen / ids.length).toFixed(2), gpt4o: gpt, win: w.toFixed(2), ok });
  }
  const totalWin = totalGpt / totalTokens;
  const avgCpt = totalAicl / totalTokens;
  results.push({ numMerges, actual: vocab.numMerges, avgCpt, totalWin, wins, trainMs, rows });
  console.log(`  actual: ${vocab.numMerges}, CPT: ${avgCpt.toFixed(2)}, win: ${totalWin.toFixed(2)}x, ${wins}/8, ${trainMs}ms`);
}

console.log('\n' + '='.repeat(85));
console.log('SWEEP RESULTS');
console.log('='.repeat(85));
console.log(`${'Req'.padStart(5)} | ${'Act'.padStart(5)} | ${'CPT'.padStart(6)} | ${'Win×'.padStart(7)} | ${'W/L'.padStart(5)} | ${'ms'.padStart(7)}`);
console.log('-'.repeat(85));
for (const r of results) {
  console.log(`${String(r.numMerges).padStart(5)} | ${String(r.actual).padStart(5)} | ${r.avgCpt.toFixed(2).padStart(6)} | ${r.totalWin.toFixed(2).padStart(7)} | ${(r.wins+'/8').padStart(5)} | ${String(r.trainMs).padStart(7)}`);
}

const best = results.reduce((a, b) => b.totalWin > a.totalWin ? b : a);
console.log(`\nBest: ${best.numMerges} merges (${best.actual} actual) → ${best.totalWin.toFixed(2)}x total, CPT ${best.avgCpt.toFixed(2)}`);
console.log(`\n${'Test'.padStart(16)} | ${'AICL'.padStart(5)} | ${'Tok'.padStart(5)} | ${'CPT'.padStart(5)} | ${'GPT-4o'.padStart(6)} | ${'Win×'.padStart(6)}`);
console.log('-'.repeat(70));
for (const r of best.rows) console.log(`${r.name.padStart(16)} | ${String(r.aicl).padStart(5)} | ${String(r.tok).padStart(5)} | ${r.cpt.padStart(5)} | ${String(r.gpt4o).padStart(6)} | ${r.win.padStart(6)}`);

File: scripts/sweep_tokenizer.py
#!/usr/bin/env python3
"""Sweep tokenizer merge counts on a compact, boundary-preserving corpus."""

from __future__ import annotations

import argparse
import sys
import time
from pathlib import Path
from typing import Any, Sequence

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

try:
    from .eval_vocab import _result_output, _tokenizer_api, evaluate, vocab_summary  # noqa: E402
    from .heldout_probes import BENCH  # noqa: E402
except ImportError:
    from eval_vocab import _result_output, _tokenizer_api, evaluate, vocab_summary  # noqa: E402
    from heldout_probes import BENCH  # noqa: E402

MERGE_COUNTS = (128, 256, 384, 512)


def _train_api() -> Any:
    try:
        from aicl.tokenizer import train_tokenizer
    except ImportError:
        from aicl import train_tokenizer  # type: ignore[attr-defined]
    return train_tokenizer


def self_check() -> None:
    assert MERGE_COUNTS == (128, 256, 384, 512)
    print("sweep_tokenizer self-check: ok")


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--corpus", type=Path, default=ROOT / "corpus" / "aicl_train.txt")
    parser.add_argument("--self-check", action="store_true")
    args = parser.parse_args(argv)
    if args.self_check:
        self_check()
        return 0
    raw = args.corpus.read_text(encoding="utf-8")
    lines = raw.split("\n")
    train_corpus = "\n".join(line for index, line in enumerate(lines) if index % 25 == 0)
    print(f"Train corpus: {len(train_corpus)} chars, {len(train_corpus.split(chr(10)))} lines")
    train_tokenizer = _train_api()
    _load, tokenize, detokenize = _tokenizer_api()
    from aicl import decode, encode

    encoded_tests = [(name, _result_output(encode(text))) for name, text in BENCH]
    results: list[dict[str, Any]] = []
    for requested in MERGE_COUNTS:
        print(f"\nTraining {requested} merges...")
        started = time.perf_counter()
        vocab = train_tokenizer(
            [train_corpus],
            {
                "num_merges": requested,
                "merge_base": 100_000,
                "max_token_length": 5,
                "min_frequency": 2,
            },
        )
        elapsed = (time.perf_counter() - started) * 1000
        rows: list[dict[str, Any]] = []
        total_aicl = 0
        total_tokens = 0
        lossless = True
        for name, raw_text in BENCH:
            aicl_text = dict(encoded_tests)[name]
            ids = tokenize(aicl_text, vocab)
            restored = detokenize(ids, vocab)
            if isinstance(restored, dict):
                restored = restored.get("output", "")
            restored = _result_output(decode(restored))
            ok = restored == raw_text
            lossless = lossless and ok
            total_aicl += len(aicl_text)
            total_tokens += len(ids)
            rows.append(
                {
                    "name": name,
                    "aicl": len(aicl_text),
                    "tokens": len(ids),
                    "cpt": len(aicl_text) / len(ids) if ids else 0.0,
                    "roundtrip": ok,
                }
            )
        result = {
            "requested": requested,
            "actual": vocab_summary(vocab)[0],
            "cpt": total_aicl / total_tokens if total_tokens else 0.0,
            "tokens": total_tokens,
            "lossless": lossless,
            "ms": elapsed,
            "rows": rows,
        }
        results.append(result)
        print(
            f"  actual: {result['actual']}, CPT: {result['cpt']:.2f}, "
            f"{total_tokens} tokens, lossless={lossless}, {elapsed:.0f}ms"
        )
    best = min(results, key=lambda result: result["tokens"])
    print("\nSWEEP RESULTS")
    print(f"{'Req':>5} | {'Act':>5} | {'CPT':>6} | {'Tok':>7} | {'ms':>9}")
    for result in results:
        print(
            f"{result['requested']:>5} | {result['actual']:>5} | {result['cpt']:>6.2f} | "
            f"{result['tokens']:>7} | {result['ms']:>9.0f}"
        )
    print(f"\nBest: {best['requested']} merges ({best['actual']} actual) -> {best['tokens']} tokens, CPT {best['cpt']:.2f}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
